staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package envelope transforms values for storage at rest using a Envelope provider
+package envelope
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	// Supportied protocol schema for gRPC.
+	tcpProtocol  = "tcp"
+	unixProtocol = "unix"
+
+	// Current version for the protocal interface definition.
+	version = "v1beta1"
+)
+
+type gRPCService struct {
+	// gRPC client instance
+	kmsClient  KmsServiceClient
+	connection *grpc.ClientConn
+}
+
+// NewEnvelopeService returns an envelope.Service which use gRPC to communicate the remote KMS provider.
+func NewEnvelopeService(endpoint, serverCert, clientCert, clientKey string) (Service, error) {
```

Comment:
Perhaps renaming this to `NewGRPCService` or `NewRemoteKMSService` might be more informative. The current name sounds a bit too generic IMO.

Reply:
Okay, will modify it to "NewGRPCService".